### PR TITLE
migration: Add progress API

### PR DESCRIPTION
**Pull request by morgo**
## A Pull Request should be associated with an Issue.

Fixes https://github.com/cashapp/spirit/issues/279

### CODE REVIEW

File: pkg/migration/runner_test.go [hunk 1]
```diff
@@ -2324,9 +2329,19 @@ func TestDeferCutOver(t *testing.T) {
 		DeferCutOver:         true,
 	})
 	assert.NoError(t, err)
-	err = m.Run(context.Background())
-	assert.Error(t, err)
-	assert.ErrorContains(t, err, "timed out waiting for sentinel table to be dropped")
+	wg := sync.WaitGroup{}
+	wg.Add(1)
```
Comment (daniel-nichter) [inline pkg/migration/runner_test.go:2333]: Just a test, but standard way to doing this:
```go
go f() {
    wg.Add(1)
    defer wg.Done()
}
```
In a test, no big deal (except maybe if `m.Run` panics, then `wg.Done` isn't called).

File: pkg/migration/runner_test.go [hunk 2]
```diff
@@ -1224,6 +1225,7 @@ func TestE2EBinlogSubscribingCompositeKey(t *testing.T) {
 	assert.NotNil(t, chunk)
 	assert.Equal(t, "((`id1` < 1001)\n OR (`id1` = 1001 AND `id2` < 1))", chunk.String())
 	assert.NoError(t, m.copier.CopyChunk(context.TODO(), chunk))
+	assert.Equal(t, Progress{CurrentState: "copyRows", Summary: "1000/1200 83.33% copyRows ETA TBD"}, m.GetProgress())
```
Comment (daniel-nichter) [inline pkg/migration/runner_test.go]: In my experience, tests on raw text output are extremely fragile and become problematic. One character changes, the tests fail, but a future engineer won't know if it's just a text diff or if there's a real issue. If there's a different way to check this, it'd be worth the time/effort to do.

File: pkg/migration/runner.go [hunk 3]
```diff
@@ -599,6 +607,29 @@ func (r *Runner) createCheckpointTable(ctx context.Context) error {
 	return nil
 }
 
+func (r *Runner) GetProgress() Progress {
+	var summary string
+	switch r.getCurrentState() { //nolint: exhaustive
```
Comment (daniel-nichter) [inline pkg/migration/runner.go:612]: Add a default to check the inevitable future bug where, somehow, it's none of the proper cases. That way, user doesn't get an empty string; they can at least get something slightly better like "Error: status unknown (report this as a bug )".

File: pkg/migration/migration.go [hunk 4]
```diff
@@ -33,14 +32,8 @@ func (m *Migration) Run() error {
 		return err
 	}
 	defer migration.Close()
-	if !m.SkipPreRunChecks {
```
Comment (daniel-nichter) [inline pkg/migration/migration.go:36]: This change doesn't seem related to adding progress?

**Review by daniel-nichter — APPROVED** [review]

File: pkg/migration/runner_test.go (hunk 1 above)
Comment (morgo) [reply pkg/migration/runner_test.go:2333]: There's a race in this suggestion though. It could reach wg.Wait() before it reaches wg.Add(1).

File: pkg/migration/migration.go (hunk 4 above)
Comment (morgo) [reply pkg/migration/migration.go:36]: No, it's removing 5.7 specific dead code I caught. See the comment just below.

File: pkg/migration/runner.go (hunk 3 above)
Comment (morgo) [reply pkg/migration/runner.go:612]: This is just for the field `summary`, which is a text-based editorialization of the CurrentState (i.e. 99% done). The `CurrentState` will always be available for the other states, but they won't have a text-based summary (for example they can't easily report progress). I don't consider this is an error, it just means there's no second level detail.

File: pkg/migration/runner_test.go (hunk 2 above)
Comment (morgo) [reply pkg/migration/runner_test.go]: There are constants for the `CurrentState`. I can always change that. There are no constants for the summary though.

File: pkg/migration/runner_test.go (hunk 1 above)
Comment (morgo) [reply pkg/migration/runner_test.go:2333]: The linter also detects the race. So I will change the defer line, but not the `wg.Add()`
